html/modules/custom/reliefweb_subscriptions/src/ReliefwebSubscriptionsMailer.php

```diff
@@ -0,0 +1,2142 @@
+          if (($length -= mb_strlen($parts[$i])) <= 0) {
+            // Truncate the paragraph and add an ellipsis.
+            $paragraphs[$index] = trim(implode(array_slice($parts, 0, $i)), $end_marks) . '...';
+            // Truncate the list of paragraphs.
+            $paragraphs = array_slice($paragraphs, 0, $index + 1);
+            // Break from both loops.
+            break 2;
+          }
+        }
+
+        // Adjust the length to reflect the added space separator when
+        // returning the text as plain text.
+        $length -= $delta;
+      }
+    }
+
+    return $prefix . implode($separator, $paragraphs) . $suffix;
+  }
+
+  /**
+   * Generate a preview.
+   */
+  public function generatePreview($sid) {
+    $subscriptions = reliefweb_subscriptions_subscriptions();
+    $subscription = $subscriptions[$sid];
+
+    if ($subscription['type'] === 'scheduled') {
+      $last = $this->getPreviousSendingTime($subscription);
+      $notification = [
+        'sid' => $sid,
+        'last' => $last,
+      ];
+      $data = $this->getScheduledNotificationData((object) $notification, $subscription);
+    }
+    else {
+      $payload = $subscription['payload'];
+      $payload['sort'] = ['id:desc'];
+      $payload['offset'] = 0;
+      $payload['limit'] = 1;
+
+      // Retrieve the data from the API.
+      $result = $this->query([], $subscription, $payload);
+      if (!empty($result) && !empty($result['data'][0]['fields'])) {
+        $data = $result['data'][0]['fields'];
+      }
+    }
+
+    // Get the mail subject.
+    $subject = $this->generateEmailSubject($subscription, $data);
+    if (empty($subject)) {
+      return '';
+    }
+
+    // Generate the HTML and text content.
+    $body = $this->generateEmailContent($subscription, $data);
+    if (empty($body)) {
+      return '';
+    }
+
+    // Render the email using the default frontend theme so that template
+    // overrides, if any, can be used.
+    $active_theme = $this->themeManager->getActiveTheme();
+    $default_theme_name = $this->themeHandler->getDefault();
+    $default_theme = $this->themeInitialization->getActiveThemeByName($default_theme_name);
+
+    $render_array = [
+      '#theme' => 'mimemail_message',
+      '#module' => 'reliefweb_subscriptions',
+      '#key' => 'notifications',
+      '#recipient' => '',
+      '#subject' => $subject,
+      '#body' => $body,
+
+    ];
+
+    $this->themeManager->setActiveTheme($default_theme);
+    $html = $this->renderer->renderRoot($render_array);
+    $this->themeManager->setActiveTheme($active_theme);
+
+    return [
+      'subject' => $subject,
+      'body' => $html,
+    ];
+  }
+
+  /**
+   * Trigger for sitreps.
+   */
+  protected function triggerOchaSitrepNotification($entity, $entity_type) {
+    if (!isset($entity->type) || $entity->type !== 'report') {
+      return FALSE;
+    }
+
+    $language = $entity->language ?? Language::LANGCODE_NOT_SPECIFIED;
```

Comment:
Not needed, the fields below (field_source, field_content_format) should be accessed the D9 way.